File: photoplethysmography/livesosfilter.py
import numpy as np
from livefilter import LiveFilter

# CLASS TAKEN FROM SAMUEL PRÖLL

class LiveSosFilter(LiveFilter):
    """Live implementation of digital filter with second-order sections.
    """
    def __init__(self, sos):
        """Initialize live second-order sections filter.

        Args:
            sos (array-like): second-order sections obtained from scipy
                filter design (with output="sos").
        """
        self.sos = sos

        self.n_sections = sos.shape[0]
        self.state = np.zeros((self.n_sections, 2))
        
    def _process(self, x):
        """Filter incoming data with cascaded second-order sections.
        """
        for s in range(self.n_sections):  # apply filter sections in sequence
            b0, b1, b2, a0, a1, a2 = self.sos[s, :]

            # compute difference equations of transposed direct form II
            y = b0*x + self.state[s, 0]
            self.state[s, 0] = b1*x - a1*y + self.state[s, 1]
            self.state[s, 1] = b2*x - a2*y
            x = y  # set biquad output as input of next filter section.

        return y

File: src/main/java/com/example/rppgkotlin/LiveSosFilter.kt
package com.example.rppgkotlin

//class inspired by the LiveSosFilter from Samuel Pröll's yarppg project
class LiveSosFilter(private val sos: Array<DoubleArray>) {
    private val nSections = sos.size
    private val state = Array(nSections) {
        DoubleArray(2)
    }

    fun process(x0: Double): Double {
        var x = x0
        var y = 0.0
        for(s in 0 until nSections){
            val b0 = sos[s][0]
            val b1 = sos[s][1]
            val b2 = sos[s][2]
            val a0 = sos[s][3]
            val a1 = sos[s][4]
            val a2 = sos[s][5]

            y = b0 * x + state[s][0]
            state[s][0] = b1 * x - a1 * y + state[s][1]
            state[s][1] = b2 * x - a2 * y
            x = y
        }
        return y
    }
}